design/one-pager-change-logs.md

```diff
@@ -0,0 +1,310 @@
+  this new `crossplane-runtime` logic and with additional code generation logic
+  to generate providers that:
+  * Create a initialize a sidecar container in the provider pod
+  * Start and initialize the gRPC server in the sidecar container and the gRPC
+    client in the main provider container
+* The [Upjet provider
+  template](https://github.com/crossplane/upjet-provider-template) is updated to
+  use this new Upjet framework version, which enables new providers to have
+  change log functionality
+* The major upjet based providers (e.g. `provider-upjet-aws`,
+  `provider-upjet-gcp`, `provider-upjet-azure`, etc.) are updated to use this
+  new version of the Upjet framework, which will generate a provider capable of
+  change log generation in the very next build/release of the provider.
+
+### Assumptions
+
+Assumptions made for this proposal include:
+
+* Managed resources throughout the Crossplane ecosystem implement full
+  `spec.forProvider` and `status.atProvider` sections so that complete desired
+  and observed state is included in the change log entries simply by serializing
+  the full resource.
+
+## Future Work
+
+The below deliverables will not be included in the initial implementation of
+this feature, but should be considered in future milestones:
+
+* The [template repo](https://github.com/crossplane/provider-template) for
+  classic providers (not generated with Upjet) should be updated to include a
+  sidecar container, gRPC connection, etc.
+* A configuration flag for Providers that allows the user to directly send the
+  change log entries to an alternate destination of their choosing besides the
+  sidecar container via gRPC connection.  This alternative would also need to
+  include specifying certificates and setting up mTLS to secure the connection.
+* Tracing of a change log entry all the way from a Claim down to a cloud API
+  call, as described in
+  https://github.com/crossplane/crossplane/issues/5477#issuecomment-1996303810.
+
+## Alternatives Considered
+
+The follow alternative approaches have been considered for this change logs design:
+
+### Kubernetes Events
+
+We could generate standard Kubernetes events with that capture all of the data
+we are proposing to store for each change log entry, likely stored as
+annotations on the events. However, given that we propose to store the entire
+object before and after a given operation is performed, this would likely be too
+large for what a standard `Event` and its annotations, or the underlying etcd
+storage, would allow. Relying on annotations also would diminish the portability
+and ease of access of the data to a wide variety of tools.
+
+### Kubernetes Audit Logs
+
+The Kubernetes [audit
+logs](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/) capture a
+similar concept to the change logs feature we are proposing, but are not a great
+fit for our goals, both in the data that it captures and the scenarios for which
```

Comment:
```suggestion
fit for our goals, both in the data that it captures and the scenarios for which
```

I don't understand this statement. The data is the same, with the exception of the optional extra JSON blob.

Reply:
Capturing data directly related to provider actions with more control about the when and why certainly has value. So there is an advantage of the controller-runtime based approach. Am just saying here that the actual data captured (modulo the additional provider-specific blob) is the same.